From dfc71d4b23d02946d28419e3522b85d84c859b05 Mon Sep 17 00:00:00 2001
From: Charles Keepax <ckeepax@opensource.wolfsonmicro.com>
Date: Mon, 2 Dec 2013 15:28:09 +0000
Subject: [PATCH 058/351] mfd: arizona: Do hard instead of soft reset coming
 out of sleep

If possible to a hardware reset instead of a software reset when coming
out of sleep, to ensure consistent register state.

Change-Id: I6a39d191c065055229b4d7fba627468052fc87f2
Signed-off-by: Charles Keepax <ckeepax@opensource.wolfsonmicro.com>
Signed-off-by: Nikesh Oswal <Nikesh.Oswal@wolfsonmicro.com>
---
 drivers/mfd/arizona-core.c |   21 ++++++++++++++++++---
 1 file changed, 18 insertions(+), 3 deletions(-)

diff --git a/drivers/mfd/arizona-core.c b/drivers/mfd/arizona-core.c
index 359b8d8..e9554f7 100644
--- a/drivers/mfd/arizona-core.c
+++ b/drivers/mfd/arizona-core.c
@@ -353,6 +353,16 @@ static int arizona_runtime_resume(struct device *dev)
 
 	dev_dbg(arizona->dev, "Leaving AoD mode\n");
 
+	switch (arizona->type) {
+	case WM5110:
+	case WM8280:
+		if (arizona->rev == 3 && arizona->pdata.reset)
+			gpio_set_value_cansleep(arizona->pdata.reset, 0);
+		break;
+	default:
+		break;
+	};
+
 	ret = regulator_enable(arizona->dcvdd);
 	if (ret != 0) {
 		dev_err(arizona->dev, "Failed to enable DCVDD: %d\n", ret);
@@ -365,9 +375,14 @@ static int arizona_runtime_resume(struct device *dev)
 	case WM5110:
 	case WM8280:
 		if (arizona->rev == 3) {
-			ret = arizona_soft_reset(arizona);
-			if (ret != 0)
-				goto err;
+			if (!arizona->pdata.reset) {
+				ret = arizona_soft_reset(arizona);
+				if (ret != 0)
+					goto err;
+			} else {
+				gpio_set_value_cansleep(arizona->pdata.reset, 1);
+				msleep(1);
+			}
 		}
 		break;
 	default:
-- 
1.7.9.5

